driver-async/src/main/com/mongodb/async/client/gridfs/GridFSDownloadStreamImpl.java

```diff
@@ -122,61 +122,87 @@ public void onResult(final GridFSFile result, final Throwable t) {
-            getBuffer(chunkIndex, new SingleResultCallback<byte[]>() {
+        } else if (!resultsQueue.isEmpty()) {
+            processResults(amountRead, dst, callback);
+        } else if (cursor == null) {
+            chunksCollection.find(new Document("files_id", fileInfo.getId())
+                    .append("n", new Document("$gte", chunkIndex)))
+                    .batchSize(batchSize).sort(new Document("n", 1))
+                    .batchCursor(new SingleResultCallback<AsyncBatchCursor<Document>>() {
+                        @Override
+                        public void onResult(final AsyncBatchCursor<Document> result, final Throwable t) {
+                            if (t != null) {
+                                callback.onResult(null, t);
+                            } else if (result == null) {
+                                chunkNotFound(chunkIndex, callback);
+                            } else {
+                                cursor = result;
+                                checkAndFetchResults(amountRead, dst, callback);
+                            }
+                        }
+                    });
+        } else {
+            cursor.next(new SingleResultCallback<List<Document>>() {
                 @Override
-                public void onResult(final byte[] result, final Throwable t) {
+                public void onResult(final List<Document> result, final Throwable t) {
                     if (t != null) {
-                        releaseReadingLock();
                         callback.onResult(null, t);
+                    } else if (result == null || result.isEmpty()) {
+                        chunkNotFound(chunkIndex, callback);
                     } else {
-                        buffer = result;
-                        readFromBuffer(dst, recursiveCallback);
+                        resultsQueue.addAll(result);
+                        if (batchSize == 1) {
+                            discardCursor();
+                        }
+                        processResults(amountRead, dst, callback);
                     }
                 }
             });
-        } else {
-            readFromBuffer(dst, recursiveCallback);
         }
     }
 
-    private void readFromBuffer(final ByteBuffer dst, final SingleResultCallback<Void> callback) {
-        int amountToCopy = dst.remaining();
-        if (amountToCopy > buffer.length - bufferOffset) {
-            amountToCopy = buffer.length - bufferOffset;
+    private void processResults(final int previousAmountRead, final ByteBuffer dst, final SingleResultCallback<Integer> callback) {
+        ChunkDataOrError chunkOrError = new ChunkDataOrError();
+        int amountRead = previousAmountRead;
+        while (currentPosition < fileInfo.getLength() && dst.remaining() > 0 && !resultsQueue.isEmpty()) {
+            if (buffer == null || bufferOffset == buffer.length) {
+                getBufferFromChunk(resultsQueue.poll(), chunkIndex, chunkOrError);
```

Comment:
I refactored, but still catch the exceptions because they aren't raised, rather they are passed to a
the users callback.  I can refactor to try catch if you feel that would improve the code.